File: android/src/main/kotlin/de/buseslaar/tracking/activity_tracking/model/Activity.kt
package de.buseslaar.tracking.activity_tracking.model

class Activity {

    // TODO: Change type to Health Connect types
    var type: ActivityType? = null

    var steps: Int = 0

    var locations: HashMap<Long, Location> = HashMap()

    var startDateTime: Long? = null;

    var endDateTime: Long? = null;

    var distance: Double = 0.0

    constructor(activityType: ActivityType) {
        type = activityType
        steps = 0
        startDateTime = System.currentTimeMillis()
    }

    fun addLocation(millis: Long, location: Location) {
        locations.put(millis, location)
    }

    fun parseToJSON(): String {
        var json = """
            {
                "startDateTime": $startDateTime,
                "endDateTime": $endDateTime,
                "type": "$type",
                "distance": $distance,
                "steps": $steps,
                "locations": {${parseHashMapToJson()}}}
        """.trimIndent()
        return json
    }

    private fun parseHashMapToJson(): String {
        var json = ""
        if (locations.isEmpty()) return json
        locations.forEach { it ->
            json =
                json + """"${it.key}": ${it.value.parseToJSON()}${if (it.key != locations.keys.last()) "," else ""}"""
        }
        return json
    }
}

File: lib/model/Activity.dart
import 'dart:ffi';

import 'package:activity_tracking/model/Location.dart';
import 'package:activity_tracking/model/activity_type.dart';

class Activity {
  // TODO: Ändere den Typ in Health Connect-Typen
  ActivityType? activityType;
  int? steps = 0;
  Map<DateTime, Location>? locations = {};
  int? startDateTime = 0;
  int? endDateTime = 0;
  double? distance = 0.0;

  Activity(
      {this.activityType,
      this.steps,
      this.locations,
      this.startDateTime,
      this.endDateTime,
      this.distance});

  factory Activity.fromJson(Map<String, dynamic> json) {
    var rawLocations = json["locations"] as Map<String, dynamic>;
    var locations = <DateTime, Location>{};
    rawLocations.forEach((key, value) {
      locations[DateTime.fromMillisecondsSinceEpoch(int.parse(key))] = Location(
          latitude: value["latitude"],
          longitude: value["longitude"],
          altitude: value["altitude"],
          speed: value["speed"]);
    });
    return Activity(
        startDateTime: json["startDateTime"],
        endDateTime: json["endDateTime"],
        activityType: json["type"],
        distance: json["distance"],
        locations: locations,
        steps: json["steps"]);
  }
}
